File: backend/app/schemas/onboarding.py
"""Onboarding questionnaire + estimate schemas (docs/API-DESIGN.md §10)."""

from __future__ import annotations

from typing import Literal

from app.schemas.common import CamelModel

# answer values are either a chosen option key (str) or a number (int)
AnswerValue = int | str


class QuestionOption(CamelModel):
    value: str
    label: str


class VisibleIf(CamelModel):
    """Show the question only when another answer satisfies this condition.

    Exactly one of equals / not_equals / any_of is set. Serializes camelCase
    (`questionId`, `notEquals`, `anyOf`) so the React renderer and the Python
    estimator evaluate identical rules.
    """

    question_id: str
    equals: AnswerValue | None = None
    not_equals: AnswerValue | None = None
    any_of: list[AnswerValue] | None = None


class Question(CamelModel):
    id: str
    type: Literal["single", "number"]
    label: str
    options: list[QuestionOption] | None = None
    min: int | None = None
    max: int | None = None
    step: int | None = None
    unit: str | None = None
    help: str | None = None
    visible_if: VisibleIf | None = None
    # R0: value-of-information 0..1 — how much answering this shrinks footprint
    # uncertainty (the questions are ordered by it, highest-yield first).
    voi: float | None = None
    default: AnswerValue


class Questionnaire(CamelModel):
    version: int
    questions: list[Question]


class OnboardingAnswers(CamelModel):
    """Submitted answers, keyed by question id. Missing keys fall back to defaults."""

    answers: dict[str, AnswerValue]


class OnboardingProgress(CamelModel):
    """Autosaved partial progress — answers so far + the step the user is on."""

    answers: dict[str, AnswerValue]
    current_step: int = 0


class OnboardingProfileOut(CamelModel):
    status: Literal["in_progress", "completed"]
    completed: bool
    current_step: int
    answers: dict[str, AnswerValue] | None = None

File: frontend/src/lib/types/onboarding.ts
/** Onboarding-questionnaire domain types (docs/API-DESIGN.md §10). */

export type QuestionType = "single" | "number";

export interface QuestionOption {
  value: string;
  label: string;
}

/** Show a question only when another answer satisfies this condition. */
export interface VisibleIf {
  questionId: string;
  equals?: AnswerValue;
  notEquals?: AnswerValue;
  anyOf?: AnswerValue[];
}

export interface Question {
  id: string;
  type: QuestionType;
  label: string;
  /** present for `single` questions. */
  options?: QuestionOption[];
  /** present for `number` questions. */
  min?: number;
  max?: number;
  step?: number;
  unit?: string;
  /** optional helper text shown under the label. */
  help?: string;
  /** gates visibility on another answer (e.g. car km only if carType ≠ none). */
  visibleIf?: VisibleIf;
  /** R0: value-of-information 0..1 — how much answering shrinks footprint uncertainty. */
  voi?: number;
  default: AnswerValue;
}

export interface Questionnaire {
  version: number;
  questions: Question[];
}

/** A single-choice option key (string) or a numeric answer. */
export type AnswerValue = number | string;

/** Answers keyed by question id. */
export type OnboardingAnswers = Record<string, AnswerValue>;

export type OnboardingStatus = "in_progress" | "completed";

export interface OnboardingProfile {
  status: OnboardingStatus;
  completed: boolean;
  /** the step the user was on — used to resume a half-finished onboarding. */
  currentStep: number;
  answers: OnboardingAnswers | null;
}
